### PR TITLE
Handle unrecognized response error code

### PR DESCRIPTION
**Pull request by chenmoneygithub**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/chenmoneygithub/mlflow/pull/10918?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10918/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10918
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Currently we assume every response should have a error code defined in `databricks.proto`, however it's possible that the request fails without reaching MLflow service, e.g., IP blacklist. In this scenario, we shouldn't throw a vague ValueError, but instead we can report the error code as it is to an `MlflowException`.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/exceptions.py [hunk 1]
```diff
@@ -119,7 +122,16 @@ def __init__(self, json):
...
-        super().__init__(message, error_code=ErrorCode.Value(error_code))
+        try:
+            super().__init__(message, error_code=ErrorCode.Value(error_code))
+        except ValueError:
+            _logger.warning(
+                f"Received error code not recognized by MLflow: {error_code}, this may indicate "
+                "your request hits error outside the MLflow server, e.g., proxy server, IP blocker "
+                "or so."
+            )
+            super().__init__(message)
```
Comment (dbczumar) [inline mlflow/exceptions.py:133]: For Databricks 403s, we *can* recognize the error code by mapping "permission_denied" to 403 or vice versa, right? Can we do that here?

Comment (chenmoneygithub) [reply mlflow/exceptions.py:133]: I don't know if we want to handle `403` specially - it's possible that other error codes are returned such as `404` or `429`. So I feel it's cleaner to keep the original message if no enum is mapped.

An alternative is to create a reverse map to [ERROR_CODE_TO_HTTP_STATUS](https://github.com/mlflow/mlflow/blob/5d13d6ec620a02de9a5e31201bf1becdb9722ea5/mlflow/exceptions.py#L28), which i feel is not quite necessary.

File: mlflow/exceptions.py [hunk 2]
```diff
@@ -114,13 +117,29 @@ class RestException(MlflowException):
...
+        except ValueError:
+            try:
+                # The `error_code` can be an http error code, in which case we convert it to the
+                # corresponding `ErrorCode`.
+                error_code = HTTP_STATUS_TO_ERROR_CODE[int(error_code)]
+                super().__init__(message, error_code=ErrorCode.Value(error_code))
+            except ValueError or KeyError:
+                _logger.warning(
+                    f"Received error code not recognized by MLflow: {error_code}, this may "
+                    "indicate your request hits error outside the MLflow server, e.g., proxy "
```
Comment (dbczumar) [inline mlflow/exceptions.py:139]:
```suggestion
                    "indicate your request encountered an error before reaching MLflow server, e.g., within a proxy "
```

File: mlflow/exceptions.py [hunk 3]
```diff
@@ -114,13 +117,29 @@ class RestException(MlflowException):
...
+                    "server, IP blocker or so."
```
Comment (dbczumar) [inline mlflow/exceptions.py:140]:
```suggestion
                    "server or authentication / authorization layer.”
```

**Review by dbczumar — APPROVED** [review]
LGTM with nit comment about logger message. Feel free to merge once addressed. Thanks @chenmoneygithub !

**Pull request merged**